### PR TITLE
fix: guess charset naively

### LINKED ISSUE #76: Content-type response header incorrect for .wasm
I've just started using sfz 0.6.1 for some WebAssembly development and I've noticed some odd behaviour.

When I request a `.wasm` file, sfz responds with `content-type: application/wasm; charset=utf-8` but this isn't a utf-8 encoded file, it's a binary format. As a result when I try to load the WebAssembly in Firefox 92, it complains about the MIME type not being as expected and having to fallback to a slower instantiation method as a result.

```text
`WebAssembly.instantiateStreaming` failed because your server does not serve wasm with `application/wasm` MIME type. Falling back to `WebAssembly.instantiate` which is slower. Original error:
 TypeError: WebAssembly: Response has unsupported MIME type 'application/wasm; charset=utf-8' expected 'application/wasm'
```

Here's a `curl -v` with sfz serving the file

```shell
*   Trying 127.0.0.1:5000...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 5000 (#0)
> GET /pkg/web_rust_bg.wasm HTTP/1.1
> Host: 127.0.0.1:5000
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< server: sfz/0.6.1
< cache-control: public, max-age=0
< last-modified: Sat, 02 Oct 2021 14:34:33 GMT
< etag: "1633185273-23174"
< accept-ranges: bytes
< content-type: application/wasm; charset=utf-8
< content-length: 23174
< date: Sat, 02 Oct 2021 19:10:56 GMT
< 
```

For extra context, here's a `curl -v` with `python3 -m http.server` serving the file, which does not produce the error

```shell
*   Trying 127.0.0.1:8000...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
> GET /pkg/web_rust_bg.wasm HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/7.68.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
* HTTP 1.0, assume close after body
< HTTP/1.0 200 OK
< Server: SimpleHTTP/0.6 Python/3.8.10
< Date: Sat, 02 Oct 2021 19:11:38 GMT
< Content-type: application/wasm
< Content-Length: 23174
< Last-Modified: Sat, 02 Oct 2021 14:34:33 GMT
< 
```

### PR DESCRIPTION
**Pull request by weihanglo**
Fixes #76.

This bug was introduced by #69. In order not to perform any additional I/O, we guess charset naively from MIME. Mainly

- `text/*`, `*/xml`, `*/javascript`, `*/json` -> UTF-8
- `*/*+xml`, `*/*+json` -> UTF-8
- others -> leave it as is

### CODE REVIEW

File: src/extensions.rs [hunk 1]
```diff
@@ -91,23 +91,47 @@ impl SystemTimeExt for SystemTime {
 
 pub trait MimeExt {
     fn is_compressed_format(&self) -> bool;
+    fn guess_chartset(&self) -> Option<mime::Name>;
```
Comment (ZacJW) [inline src/extensions.rs:94]: Typo here, should be `guess_charset` not `guess_chartset`. This will need to be changed for all references too.

**Comment by ZacJW** [comment]
Just compiled this and this change does fix #76, so it's fine for my usecase. Can't help thinking that a more thorough check may later prove necessary.

As a quick test I had sfz serve a UTF-16 LE enocded .txt file and the content type header was mistakenly `text/plain; charset=utf-8`. I would argue that if sfz is not going to verify the file is of any particular charset, it should leave it out of the content-type altogether and let the client figure it out rather than lead them down the wrong path.

If reporting the charset proves necessary for some usecase, maybe sfz could check the encoding of the file (possibly using encoding_rs if introducing a dependency is acceptable) and cache the determined encoding to avoid checking every time.

Thanks for the quick fix, hope I've not made too much work for you.

**Comment by ZacJW** [comment]
I've done a bit more investigation so I'll share it here. As I understand it, your decision to default to `charset=utf-8` was informed by #68 in which they have an issue viewing yaml with non ASCII characters in chrome when served by sfz 0.6.0.  However as per [yaml 1.2.2 spec](https://yaml.org/spec/1.2.2/#52-character-encodings) (and [since 1.0](https://yaml.org/spec/1.0/index.html#id2559514)) all yaml character streams are unicode, defaulting to utf-8 encoding. This leads me to believe that the bug in #68 is acutally a bug in chrome, not sfz. I've reproduced that bug with sfz 0.6.0 and with a modified sfz 0.6.1 that never includes a charset in the content-type header. This is all viewed with chromium 94.0.4606.71.

**Comment by weihanglo** [comment]
Thanks for the review and research! Your concern is reasonable. And yes, this is a thing that client can handle. Unfortunately, the market share of Chromium is large which we cannot simply ignore this inconvenience.

My concern is that Encoding detection is hard and error-prone. It's time to introduce additional dependency to handle this. I plan to add a flag to toggle this auto-detection feature.

**Comment by ZacJW** [comment]
You're right about not being able to ignore chromium. Looks like they're still using Windows-1252 encoding for this by default :laughing:. Guess I'll go write a bug report for them. Thanks again.

**Comment by weihanglo** [comment]
I am going to merge this PR since it at least make the blind guess a bit better. 
Next, I'll try to integrate [chardetng](https://crates.io/crates/chardetng), which is used in Firefox. Its [design doc](https://hsivonen.fi/chardetng/) is very well-written and worth a read 👍🏾

Thanks for your review again!

**Pull request merged**